libs/encryption-sdk/src/main/java/org/opensearch/encryption/core/FrameEncryptionHandler.java

```diff
@@ -0,0 +1,344 @@
+    private final byte[] messageId_;
+    private final int frameSize_;
+    private final int tagLenBytes_;
+
+    private long frameNumber_ = 1;
+    private boolean isFinalFrame_;
+
+    private final byte[] bytesToFrame_;
+    private int bytesToFrameLen_;
+    private boolean complete_ = false;
+
+    /**
+     * Construct an encryption handler for encrypting bytes and storing them in frames.
+     *
+     * @param encryptionKey Encryption key used to encrypt file content. Appended to file header.
+     * @param nonceLen random number appended with every frame header.
+     * @param cryptoAlgo Crypto algorithm used to perform encryption
+     * @param messageId Message Id from the crypto header
+     * @param frameSize Size of a frame.
+     * @param frameStartNumber Number from which assignment has to start for new frames
+     */
+    public FrameEncryptionHandler(
+        final SecretKey encryptionKey,
+        final int nonceLen,
+        final CryptoAlgorithm cryptoAlgo,
+        final byte[] messageId,
+        final int frameSize,
+        final int frameStartNumber
+    ) {
+        encryptionKey_ = encryptionKey;
+        cryptoAlgo_ = cryptoAlgo;
+        nonceLen_ = nonceLen;
+        messageId_ = messageId.clone();
+        frameSize_ = frameSize;
+        tagLenBytes_ = cryptoAlgo_.getTagLen();
+        bytesToFrame_ = new byte[frameSize_];
+        bytesToFrameLen_ = 0;
+        cipherHandler_ = new CipherHandler(encryptionKey_, Cipher.ENCRYPT_MODE, cryptoAlgo_);
+        frameNumber_ = frameStartNumber;
+    }
+
+    /**
+     * Encrypt a block of bytes from in putting the plaintext result into out.
+     *
+     * <p>It encrypts by performing the following operations:
+     *
+     * <ol>
+     *   <li>determine the size of encrypted content that can fit into current frame
+     *   <li>call processBytes() of the underlying cipher to do corresponding cryptographic encryption
+     *       of plaintext
+     *   <li>check if current frame is fully filled using the processed bytes, write current frame to
+     *       the output being returned.
+     * </ol>
+     *
+     * @param in the input byte array.
+     * @param out the output buffer the encrypted bytes go into.
+     * @param outOff the offset into the output byte array the encrypted data starts at.
+     * @return the number of bytes written to out and processed
+     */
+    @Override
+    public ProcessingSummary processBytes(final byte[] in, final int off, final int len, final byte[] out, final int outOff)
+        throws BadCiphertextException {
+        int actualOutLen = 0;
+
+        int size = len;
+        int offset = off;
+        while (size > 0) {
+
+            final int currentFrameCapacity = frameSize_ - bytesToFrameLen_;
+            // bind size to the capacity of the current frame
+            size = Math.min(currentFrameCapacity, size);
+
+            System.arraycopy(in, offset, bytesToFrame_, bytesToFrameLen_, size);
```

Comment:
do we need to copy bytes here? can we just send `in` with `offset` and `size `

Reply:
It is always a good practice to avoid changing the inputs unless it is intended. Also, it is coming from encryption SDK so no custom change there.